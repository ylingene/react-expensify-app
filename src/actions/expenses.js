import uuid from 'uuid'; // create unique identifiers
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});
// export const addExpense = (
//     {
//         description = '',
//         note = '', amount = 0,
//         createdAt = 0
//     } = {}
// ) => ({
//     type: 'ADD_EXPENSE',
//     expense: {
//         id: uuid(),
//         description,
//         note,
//         amount,
//         createdAt
//     }
// });

// requires middleware Thunk
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        // added return to allow for promise chaining
        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
// I guess you don't need to have just an expense or filter obj for an action if only using a piece of info
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(expenses));
        });
    };
}