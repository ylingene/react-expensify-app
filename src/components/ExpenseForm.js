import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// reuseable component for add & edit expenses
// pass in onSubmit through props to dictate how to update the store
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: '',
        };
    }

    // track state of form locally, then use it to update redux on form submit
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = (e) => {
        const note = e.target.value;
       ; this.setState(() => ({ note }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = (createdAt) => {
        // prevent user to clear value (hit delete key on date picker focus)
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount.' }))
        }
        else {
            // Clear the error
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100, // we store as cents
                createdAt: this.state.createdAt.valueOf(), // gets back Unixtimestamp in milliseconds
                note: this.state.note
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input 
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker 
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add Expense</button>
                </form>
            </div>
        )
    }
}