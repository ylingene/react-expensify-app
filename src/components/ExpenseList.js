import React from 'react';
import { connect } from 'react-redux'; // allow connection to the redux store
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div className="content-container">
        <div className="list-header">
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
        <div className="list-body">
            {
                props.expenses.length === 0 ? (
                    <div className="list-item list-item--message">
                        <span>No Expenses</span>
                    </div>
                ) : (
                    props.expenses.map((expense, idx) => (
                        <ExpenseListItem
                            key={idx}
                            {...expense} />
                    ))
                )
            }
        </div>
    </div>
);

// define function in connect() to say what subset of info from store the component needs to access
// passed into as argument in connect()
// Note: this function has access to store state, and can define function to access data I need
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// connect() returns a function, pass component into it to create a HOC (higher order component). 
    // that function returns a component as well --> HOC
// if nothing passed into connect, will just give component access to dispatch() function
// different way to export default
export default connect(mapStateToProps)(ExpenseList);

// convenient becuse react-redux handles 'subscribing' to data changes in the store
// and the components is automatically updated w/ new data.
// only need to decide what to do with the data

// expense didn't need any data passed down, just accessed through redux store
// Question: when to use redux store vs. pass down through nested components??
