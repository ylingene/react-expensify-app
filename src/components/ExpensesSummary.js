import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expensesCount, expensesTotal }) => (
    <div>
        <p>{`Viewing ${expensesCount} expense${expensesCount === 1 ? '' : 's'} totalling ${numeral(expensesTotal / 100).format('$0,0.00')}`}</p>
    </div>
);

const mapStateToProps = (state) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesCount: expenses.length,
        expensesTotal: selectExpensesTotal(expenses)
    };
};

// const mapDispatchToProps = () => {

// }

export default connect(mapStateToProps)(ExpensesSummary);