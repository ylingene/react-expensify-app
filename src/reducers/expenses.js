// Expenses Reducer
const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // spread operator works like concat
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            // if true, keep. else, remove from array
            return state.filter(({ id }) => action.id !== id);
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    // object spread operator (need to import package transform-object-rest-spread)
                    return {
                        ...expense,
                        ...action.updates // override values in expense with those in updates
                    }
                }
                else {
                    return expense; // do nothing if not match
                }
            });
        default:
            return state;
    }
};

export default expensesReducer;