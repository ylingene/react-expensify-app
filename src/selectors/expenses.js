import moment from 'moment';

// timestamps (milliseconds) for startDate & endDate
// January 1st 1970 @ midnight (unix epoch)
// ex: 33400, 10, -203

// Get visible expenses
// Destructure filters object (second argument) into indiv variable names
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdAtMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1; // latest comes first
        }
        else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1; // highest comes first
        }
    });
};

export default getVisibleExpenses;