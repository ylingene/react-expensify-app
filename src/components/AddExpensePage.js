import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses'

// AddExpensePage refactored to use mapDispatchToProps as abstraction 
    // but also allow flexibility for testing by using a spy for addExpense

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        // props.dispatch(addExpense(expense)); // refactored
        this.props.startAddExpense(expense);

        this.props.history.push('/'); // use browser routing to redirect back to expense dashboard
        // passed in from React Router when routing from a different component
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm
                    onSubmit={this.onSubmit}
                />
            </div>
        )
    }
}

// create abstraction from component of global addExpense
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
