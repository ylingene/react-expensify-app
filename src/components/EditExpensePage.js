import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startRemoveExpense, editExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    };

    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>
                    Remove
                </button>
            </div>
        );
    };
}

// also have access to props that's going to be passed into component
const mapStateToProps = (state, props) => {
    return {
        // props.match is part of React Router
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    };
};

// create abstraction from component of global addExpense
// props argument not necessary but available if needed
const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, updates) => dispatch(editExpense(id, updates)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
