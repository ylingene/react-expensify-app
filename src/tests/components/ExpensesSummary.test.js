import React from 'react';
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary';
import selectExpensesTotal from '../../selectors/expenses-total';
import expenses from '../fixtures/expenses';

test('render ExpensesSummary with more than 1 expenses', () => {
    const expensesCount = expenses.length;
    const expensesTotal = selectExpensesTotal(expenses);
    const wrapper = shallow(<ExpensesSummary expensesCount={expensesCount} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});

test('render ExpensesSummary with 1 expense', () => {
    const expensesCount = 1;
    const expensesTotal = selectExpensesTotal([expenses[0]]);
    const wrapper = shallow(<ExpensesSummary expensesCount={expensesCount} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});

test('render ExpensesSummary with 0 expenses', () => {
    const expensesCount = 0;
    const expensesTotal = selectExpensesTotal([]);
    const wrapper = shallow(<ExpensesSummary expensesCount={expensesCount} expensesTotal={expensesTotal} />);
    expect(wrapper).toMatchSnapshot();
});