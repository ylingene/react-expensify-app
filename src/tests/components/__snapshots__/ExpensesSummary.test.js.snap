// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpensesSummary with 0 expenses 1`] = `
<div>
  <p>
    Viewing 0 expenses totalling $0.00
  </p>
</div>
`;

exports[`render ExpensesSummary with 1 expense 1`] = `
<div>
  <p>
    Viewing 1 expense totalling $1.95
  </p>
</div>
`;

exports[`render ExpensesSummary with more than 1 expenses 1`] = `
<div>
  <p>
    Viewing 3 expenses totalling $241.95
  </p>
</div>
`;
